result-watcher.bal

```diff
@@ -299,10 +301,65 @@ public isolated class ResultWatcher {
     private isolated function checkTaskResult(TaskStatusResponse result) {
         if result.status == "UNKNOWN" || result.status == "PENDING" {
             // In case of pending, check if new substep... progress and step list auslesen, vergleiche substeps... mit fehler abbrechen, wenn plugin blödsinn macht (substeps löscht), oder warning in log von step
-            // TODO check processor.processSubstep()
-            // write progress into db
-            // save current task log
-            return; // nothing to do, still waiting for result
+
+            // TODO: update progress in any case???
+            database:TimelineSubstep[]? resultSubsteps = result?.substeps;
+            database:Progress? progress = result?.progress;
+            string? taskLog = result?.taskLog;
+            if resultSubsteps != () {
+                transaction {
+                    database:TimelineSubstepSQL[] substeps = check database:getTimelineSubsteps(self.stepId);
+                    // cast db substeps to format in result
+                    database:TimelineSubstep[] tempSubsteps = [];
+                    foreach var substep in substeps {
+                        database:TimelineSubstep tempSubstep = {
+                            substepId: substep.substepId,
+                            href: substep.href,
+                            hrefUi: substep.hrefUi,
+                            cleared: substep.cleared
+                        };
+                        tempSubsteps.push(tempSubstep);
+                    }
+                    // find new/changed substeps
+                    database:TimelineSubstep[] newSubsteps = from var substep in tempSubsteps
+                        where tempSubsteps.indexOf(substep) == ()
+                        select substep;
+                    // add in db
+                    foreach var substep in newSubsteps {
+                        check database:updateTimelineSubstep(self.stepId, substep); // TODO: perhaps nicer to pass TimelineSubstep
+                    }
+
+                    // TODO: check that UI didn't delete substeps
+
+                    // write progress into db - 
+                    if progress != () && newSubsteps.length() > 0 {
```

Comment:
Progress is independent of substeps...
Progress updates may also be merged with task log updates. They concern the same database row and happen always at the same time.